external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSourceRDD.scala

```
@@ -0,0 +1,163 @@
+    val res = context.runJob(
+      this,
+      (tc: TaskContext, it: Iterator[ConsumerRecord[Array[Byte], Array[Byte]]]) =>
+      it.take(parts(tc.partitionId)).toArray, parts.keys.toArray
+    )
+    res.foreach(buf ++= _)
+    buf.toArray
+  }
+
+  override def compute(
+      thePart: Partition,
+      context: TaskContext): Iterator[ConsumerRecord[Array[Byte], Array[Byte]]] = {
+    val range = thePart.asInstanceOf[KafkaSourceRDDPartition].offsetRange
+    assert(
+      range.fromOffset <= range.untilOffset,
+      s"Beginning offset ${range.fromOffset} is after the ending offset ${range.untilOffset} " +
+        s"for topic ${range.topic} partition ${range.partition}. " +
+        "You either provided an invalid fromOffset, or the Kafka topic has been damaged")
+    if (range.fromOffset == range.untilOffset) {
+      logInfo(s"Beginning offset ${range.fromOffset} is the same as ending offset " +
+        s"skipping ${range.topic} ${range.partition}")
+      Iterator.empty
+
+    } else {
+
+      val consumer = CachedKafkaConsumer.getOrCreate(
+        range.topic, range.partition, executorKafkaParams)
+      var requestOffset = range.fromOffset
+
+      logDebug(s"Creating iterator for $range")
+
+      new Iterator[ConsumerRecord[Array[Byte], Array[Byte]]]() {
+
+        private var prefetch: ConsumerRecord[Array[Byte], Array[Byte]] = _
+
+        private def fetchNext(): ConsumerRecord[Array[Byte], Array[Byte]] = {
+          try {
+            val r = consumer.get(requestOffset)
+            requestOffset += 1
+            r
+          } catch {
+            case e: OffsetOutOfRangeException =>
+              logWarning(s"${range.topicPartition} was deleted, some data may have been missed")
+              null
+          }
+        }
+
+        override def hasNext(): Boolean = {
+          if (prefetch == null && requestOffset < range.untilOffset) {
+            prefetch = fetchNext()
+          }
+          prefetch != null
```

I'm assuming the rest of these changes and the addition of prefetch is to support silently returning the wrong thing in case of offset out of range?